### PR TITLE
Use safe_url_string in link extraction

### LINKED ISSUE #998: HTML entity causes UnicodeEncodeError in LxmlLinkExtractor
A page containing &amp;#9001; causes with error in LxmlLinkExtractor.link_extractor()

```
          File "/[path]/spiders/search_spider.py", line 112, in parse
            self.link_extractor.extract_links(response) if
          File "/[path to virtualenv]/local/lib/python2.7/site-packages/scrapy/contrib/linkextractors/lxmlhtml.py", line 107, in extract_links
            links = self._extract_links(doc, response.url, response.encoding, base_url)
          File "/[path to virtualenv]/local/lib/python2.7/site-packages/scrapy/linkextractor.py", line 94, in _extract_links
            return self.link_extractor._extract_links(*args, **kwargs)
          File "/[path to virtualenv]/local/lib/python2.7/site-packages/scrapy/contrib/linkextractors/lxmlhtml.py", line 57, in _extract_links
            url = url.encode(response_encoding)
          File "/[path to virtualenv]/lib/python2.7/encodings/cp1252.py", line 12, in encode
            return codecs.charmap_encode(input,errors,encoding_table)
        exceptions.UnicodeEncodeError: 'charmap' codec can't encode character u'\u2329' in position 87: character maps to <undefined>
```

The encoding is correct and works for displaying the page in a browser, but fails as above in Scrapy.

### LINKED ISSUE #1403: Exception in LxmLinkExtractor.extract_links 'charmap' codec can't encode character
```
Stacktrace (most recent call last):

  File "scrapy/utils/defer.py", line 102, in iter_errback
    yield next(it)
  File "scrapy/spidermiddlewares/offsite.py", line 28, in process_spider_output
    for x in result:
  File "scrapy/spidermiddlewares/referer.py", line 22, in <genexpr>
    return (_set_referer(r) for r in result or ())
  File "scrapy/spidermiddlewares/offsite.py", line 28, in process_spider_output
    for x in result:
  File "scrapy/spidermiddlewares/urllength.py", line 37, in <genexpr>
    return (r for r in result or () if _filter(r))
  File "scrapy/spidermiddlewares/depth.py", line 54, in <genexpr>
    return (r for r in result or () if _filter(r))
  File "scrapy/spiders/crawl.py", line 69, in _parse_response
    for requests_or_item in iterate_spider_output(cb_res):
  File "ex_link_crawl/spiders/external_link_spider.py", line 45, in parse_obj
    for link in LxmlLinkExtractor(allow=(), deny=self.allowed_domains).extract_links(response):
  File "scrapy/linkextractors/lxmlhtml.py", line 108, in extract_links
    links = self._extract_links(doc, response.url, response.encoding, base_url)
  File "scrapy/linkextractors/__init__.py", line 103, in _extract_links
    return self.link_extractor._extract_links(*args, **kwargs)
  File "scrapy/linkextractors/lxmlhtml.py", line 57, in _extract_links
    url = url.encode(response_encoding)
  File "encodings/cp1252.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_table)

```

My use of extractor is following:

```
def parse_obj(self, response):
        if not isinstance(response, HtmlResponse):
            return
        for link in LxmlLinkExtractor(allow=(), deny=self.allowed_domains).extract_links(response):
            if not link.nofollow:
                yield LinkCrawlItem(domain=link.url)
```

### PR DESCRIPTION
**Pull request by Gallaecio**
Fixes #1403, fixes #998 (https://github.com/scrapy/scrapy/issues/998#issuecomment-247001012), closes #1949

There are 2 test output changes:
- 1 test change is to adjust expectations to those reported at https://github.com/scrapy/scrapy/pull/1949#issuecomment-212954241 in line with the issue being fixed
- Another test change consists of link extractors encoding whitespace, something they [stopped doing in 1.4.0](https://github.com/scrapy/scrapy/commit/2b4d46315f2ce7ce20e7355fe4ad184bb041d6f4) and would be doing again now. I don’t think it is a major issue, and in fact I find it weird that everything in the path was quoted but whitespace, so it may even be a change for the better, consistency-wise.

### CODE REVIEW

File: scrapy/linkextractors/lxmlhtml.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def _extract_links(self, selector, response_url, response_encoding, base_url):
                 url = self.process_attr(attr_val)
                 if url is None:
                     continue
-            url = to_unicode(url, encoding=response_encoding)
+            url = canonicalize_url(url, encoding=response_encoding)
```
Comment (kmike) [inline scrapy/linkextractors/lxmlhtml.py]: I'm not sure this is correct; canonicalize_url is not guaranteed to be safe - it could be the case that server won't accept URL after canonicalization. We should never be downloading results of canonicalize_url; its use should be limited to duplicate checks.
